test/modules/TestResultCoordinator/Reports/EdgeHubRestartTest/EdgeHubRestartDirectMethodReportGenerator.cs

```diff
@@ -0,0 +1,303 @@
+            this.ReceiverTestResults = Preconditions.CheckNotNull(receiverTestResults, nameof(receiverTestResults));
+            this.completedStatusHistogram = new Dictionary<HttpStatusCode, List<TimeSpan>>();
+        }
+
+        internal string TrackingId { get; }
+
+        internal string SenderSource { get; }
+
+        internal string ReceiverSource { get; }
+
+        internal TestReportType TestReportType { get; }
+
+        internal ITestResultCollection<TestOperationResult> SenderTestResults { get; }
+
+        internal ITestResultCollection<TestOperationResult> ReceiverTestResults { get; }
+
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            Logger.LogInformation($"Generating report: {nameof(EdgeHubRestartDirectMethodReport)} for [{this.SenderSource}] and [{this.ReceiverSource}]");
+
+            long previousSeqNum = 0;
+            ulong passedDirectMethodCount = 0;
+            ulong senderResultCount = 0;
+            ulong receiverResultCount = 0;
+            bool hasSenderResult = true;
+            long senderSeqNum = 0;
+            bool hasReceiverResult = true;
+            long receiverSeqNum = 0;
+
+            (senderResultCount, hasSenderResult, senderSeqNum) =
+                await this.MoveNextSenderResultAsync(senderResultCount);
+
+            (receiverResultCount, hasReceiverResult, receiverSeqNum) =
+                await this.MoveNextReceiverResultAsync(receiverResultCount);
+
+            while (hasSenderResult && hasReceiverResult)
+            {
+                this.ValidateResult(
+                    this.SenderTestResults.Current,
+                    this.SenderSource,
+                    TestOperationResultType.EdgeHubRestartDirectMethod.ToString());
+
+                this.ValidateResult(
+                    this.ReceiverTestResults.Current,
+                    this.ReceiverSource,
+                    TestOperationResultType.DirectMethod.ToString());
+
+                if (receiverSeqNum > senderSeqNum)
+                {
+                    // Increment sender result to have the same seq as the receiver
+                    (senderResultCount, hasSenderResult, senderSeqNum) = await this.IncrementSequenceNumberAsync(
+                        hasSenderResult,
+                        this.MoveNextSenderResultAsync,
+                        receiverSeqNum,
+                        senderResultCount);
+                }
+
+                if (receiverSeqNum < senderSeqNum)
+                {
+                    // Increment receiver result to have the same seq as the sender
+                    (receiverResultCount, hasReceiverResult, receiverSeqNum) = await this.IncrementSequenceNumberAsync(
+                        hasReceiverResult,
+                        this.MoveNextReceiverResultAsync,
+                        senderSeqNum,
+                        receiverResultCount);
+                }
+
+                if (hasSenderResult ^ hasReceiverResult)
```

Comment:
nit
```suggestion
                if (!hasSenderResult || !hasReceiverResult)
```

Reply:
I think XOR is more readable to me.

Reply:
+1 on Philip's comment. DOn't use bit operators like XOR on boolean variables.